File: experiments/ts-api-digest/digest.py
#!/usr/bin/env python3
"""`mi api` pour TypeScript : surface publique seule, corps jetes, types
internes atteignables conserves. Compare a la strategie head/tail de dsh."""
import re, glob, os

THRESHOLD, HEAD, TAIL = 8192, 4096, 1024
MARKER = "\n\n[... tool result middle pruned ...]\n\n"
def dsh_prune(s):
    return s if len(s) <= THRESHOLD else s[:HEAD] + MARKER + s[-TAIL:]

DECL = re.compile(r'^(export\s+)?(default\s+)?(declare\s+)?(abstract\s+)?(async\s+)?'
                  r'(class|interface|namespace|enum|type|function|const|let|var)\s+'
                  r'([A-Za-z_$][\w$]*)')
DESCEND = re.compile(r'^\s*(export\s+)?(default\s+)?(declare\s+)?(abstract\s+)?'
                     r'(class|interface|namespace|enum)\b')
KEEPVAL = re.compile(r'^\s*(export\s+)?(declare\s+)?(const|let|var|type)\b')

def strip_bodies(lines):
    """Garde les signatures, jette les corps de fonctions et de methodes."""
    out, depth, skip = [], 0, None
    for ln in lines:
        o, c = ln.count('{'), ln.count('}')
        if skip is not None:
            depth += o - c
            if depth <= skip: skip = None
            continue
        net = o - c
        if net > 0 and not (DESCEND.match(ln) or KEEPVAL.match(ln)):
            out.append(re.sub(r'\s*\{\s*$', '', ln.rstrip()))
            skip = depth; depth += net; continue
        out.append(ln); depth += net
    return out

def chunks(src):
    """Decoupe le fichier en blocs de premier niveau, commentaires attaches."""
    lines, res, buf, depth, pending = src.split('\n'), [], [], 0, []
    for ln in lines:
        s = ln.strip()
        if depth == 0 and not buf and (s.startswith('//') or s.startswith('/*')
                                       or s.startswith('*') or s == ''):
            pending.append(ln); continue
        buf.append(ln)
        depth += ln.count('{') - ln.count('}')
        if depth <= 0 and buf:
            txt = '\n'.join(buf)
            if txt.strip():
                res.append((pending, buf))
            buf, pending, depth = [], [], 0
    if buf: res.append((pending, buf))
    return res

def api_digest(src, keep_doc=False):
    imports, kept, internal = [], [], []
    for pending, body in chunks(src):
        head = body[0].lstrip()
        if head.startswith('import ') or head.startswith('export * ') or head.startswith("import{"):
            m = re.search(r"from\s+'([^']+)'", '\n'.join(body))
            if m: imports.append(m.group(1).split('/')[-1])
            continue
        m = DECL.match(head)
        sig = '\n'.join(strip_bodies(body))
        doc = '\n'.join(l for l in pending if l.strip().startswith(('/**','*','*/'))) if keep_doc else ''
        entry = ((doc + '\n' if doc.strip() else '') + sig, m.group(7) if m else None)
        if head.startswith('export'):
            kept.append(entry)
        elif m and m.group(6) in ('type', 'interface', 'enum'):
            internal.append(entry)
        # le reste (helpers internes, constantes privees) est jete

    # types internes atteignables depuis la surface publique
    text = '\n'.join(t for t, _ in kept)
    changed = True
    while changed:
        changed = False
        for e in internal[:]:
            t, name = e
            if name and re.search(r'\b%s\b' % re.escape(name), text):
                kept.append(e); internal.remove(e); text += '\n' + t; changed = True

    out = []
    if imports:
        out.append('// imports: ' + ', '.join(sorted(set(imports))))
    out += [t for t, _ in kept]
    return re.sub(r'\n{3,}', '\n\n', '\n\n'.join(out)).strip() + '\n'

SYM = re.compile(r'^export\s+(?:default\s+)?(?:declare\s+)?(?:abstract\s+)?'
                 r'(?:async\s+)?(?:class|interface|type|enum|function|const|let|var)\s+'
                 r'([A-Za-z_$][\w$]*)', re.M)

rows, T = [], [0]*7
for path in sorted(glob.glob('dsh-sample/*.ts')):
    src = open(path, encoding='utf-8').read()
    if len(src) <= THRESHOLD: continue
    p, d, dd = dsh_prune(src), api_digest(src), api_digest(src, keep_doc=True)
    syms = set(SYM.findall(src))
    kp = sum(1 for s in syms if re.search(r'\b%s\b' % re.escape(s), p))
    kd = sum(1 for s in syms if re.search(r'\b%s\b' % re.escape(s), d))
    r = (os.path.basename(path).replace('core_','').replace('_src_','/'),
         len(src), len(p), len(d), len(dd), len(syms), kp, kd)
    rows.append(r)
    for i,v in enumerate(r[1:]): T[i]+=v

print(f"{'fichier':<24}{'source':>9}{'head/tail':>10}{'digest':>8}{'+doc':>8}{' sym':>5}{' h/t':>5}{' dig':>5}")
print('-'*76)
for r in rows: print(f"{r[0]:<24}{r[1]:>9,}{r[2]:>10,}{r[3]:>8,}{r[4]:>8,}{r[5]:>5}{r[6]:>5}{r[7]:>5}")
print('-'*76)
print(f"{'TOTAL':<24}{T[0]:>9,}{T[1]:>10,}{T[2]:>8,}{T[3]:>8,}{T[4]:>5}{T[5]:>5}{T[6]:>5}")
print()
print(f"head/tail   : {T[1]/T[0]*100:5.1f} % des caracteres · {T[5]/T[4]*100:5.1f} % des symboles publics")
print(f"digest      : {T[2]/T[0]*100:5.1f} % des caracteres · {T[6]/T[4]*100:5.1f} % des symboles publics")
print(f"digest +doc : {T[3]/T[0]*100:5.1f} % des caracteres · {T[6]/T[4]*100:5.1f} % des symboles publics")

File: experiments/ts-api-digest/digest.mjs
// `mi api` pour TypeScript — extraction de surface publique via l'AST du
// compilateur TypeScript (pas de regex, pas de comptage d'accolades).
import { createRequire } from 'node:module'
const ts = createRequire(import.meta.url)('typescript')

const has = (n, k) => n.modifiers?.some(m => m.kind === k) ?? false
const isExported = n => has(n, ts.SyntaxKind.ExportKeyword)
const isPrivate  = n => has(n, ts.SyntaxKind.PrivateKeyword)

function jsdoc(node, sf, src, keepDoc) {
  if (!keepDoc) return ''
  const full = node.getFullText(sf)
  const own = src.slice(node.getStart(sf), node.end)
  const lead = full.slice(0, full.length - own.length)
  const m = lead.match(/\/\*\*[\s\S]*?\*\//g)
  return m ? m[m.length - 1] + '\n' : ''
}

function memberText(m, sf, src) {
  if (isPrivate(m)) return null
  if (m.body) return src.slice(m.getStart(sf), m.body.getStart(sf)).replace(/\s*$/, '')
  if (ts.isPropertyDeclaration(m) && m.initializer && !m.type)
    return src.slice(m.getStart(sf), m.initializer.getStart(sf)).replace(/[\s=]*$/, '')
  return m.getText(sf)
}

function declText(node, sf, src) {
  if (ts.isFunctionDeclaration(node))
    return node.body ? src.slice(node.getStart(sf), node.body.getStart(sf)).replace(/\s*$/, '')
                     : node.getText(sf)

  if (ts.isClassDeclaration(node)) {
    const head = src.slice(node.getStart(sf), node.members.pos).replace(/\s*$/, '')
    const body = node.members.map(m => memberText(m, sf, src)).filter(Boolean)
      .map(t => '  ' + t.trim())
    return head + '\n' + body.join('\n') + '\n}'
  }

  if (ts.isVariableStatement(node)) {
    const parts = node.declarationList.declarations.map(d => {
      if (d.type) return src.slice(d.getStart(sf), d.type.end)
      if (d.initializer) {
        const init = d.initializer.getText(sf)
        if (init.length <= 160) return d.getText(sf)
        return src.slice(d.getStart(sf), d.initializer.getStart(sf)) + '/* … */'
      }
      return d.getText(sf)
    })
    const kw = node.declarationList.flags & ts.NodeFlags.Const ? 'const'
             : node.declarationList.flags & ts.NodeFlags.Let ? 'let' : 'var'
    return (isExported(node) ? 'export ' : '') + kw + ' ' + parts.join(', ')
  }

  return node.getText(sf)   // interface, type, enum, module…
}

const NAMED = new Set([
  ts.SyntaxKind.InterfaceDeclaration, ts.SyntaxKind.TypeAliasDeclaration,
  ts.SyntaxKind.EnumDeclaration,
])

export function apiDigest(src, { keepDoc = false, file = 'x.ts' } = {}) {
  const sf = ts.createSourceFile(file, src, ts.ScriptTarget.Latest, true)
  const imports = [], pub = [], internal = []

  for (const st of sf.statements) {
    if (ts.isImportDeclaration(st) || ts.isImportEqualsDeclaration(st)) {
      const m = st.getText(sf).match(/from\s+['"]([^'"]+)['"]/)
      if (m) imports.push(m[1].split('/').pop())
      continue
    }
    if (ts.isExportDeclaration(st)) { pub.push({ text: st.getText(sf), name: null }); continue }

    const name = st.name?.getText(sf)
      ?? st.declarationList?.declarations?.[0]?.name?.getText(sf) ?? null
    const text = jsdoc(st, sf, src, keepDoc) + declText(st, sf, src)
    const entry = { text, name }

    if (isExported(st)) pub.push(entry)
    else if (NAMED.has(st.kind)) internal.push(entry)
    // le reste (helpers privés, constantes internes) est jeté
  }

  // types internes atteignables depuis la surface publique
  let joined = pub.map(e => e.text).join('\n')
  for (let changed = true; changed; ) {
    changed = false
    for (const e of [...internal]) {
      if (e.name && new RegExp(`\\b${e.name}\\b`).test(joined)) {
        pub.push(e); internal.splice(internal.indexOf(e), 1)
        joined += '\n' + e.text; changed = true
      }
    }
  }

  const out = []
  if (imports.length) out.push('// imports: ' + [...new Set(imports)].sort().join(', '))
  out.push(...pub.map(e => e.text))
  return out.join('\n\n').replace(/\n{3,}/g, '\n\n').trim() + '\n'
}

export function publicSymbols(src, file = 'x.ts') {
  const sf = ts.createSourceFile(file, src, ts.ScriptTarget.Latest, true)
  const out = new Set()
  for (const st of sf.statements) {
    if (!isExported(st)) continue
    if (st.name) out.add(st.name.getText(sf))
    else if (st.declarationList)
      for (const d of st.declarationList.declarations) out.add(d.name.getText(sf))
  }
  return out
}
